### PR TITLE
fix(datasource/go): Fixes splitting of gitlab go package name

### PR DESCRIPTION
**Pull request by ManuelB**
Here is the corresponding discussion:
https://github.com/renovatebot/renovate/discussions/26408

## Changes

This will correctly split the registryUrl and the packageName of a go package.

## Context

It is necessary to split a typical go modul into an url and a package name for gitlab.

e.g. example.com/gitlab/my-repo.git

registryUrl: example.com/gitlab/
packageName: my-repo

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/go/base.ts [hunk 1]
```diff
@@ -154,17 +155,35 @@ export class BaseGoDatasource {
       const parsedUrl = URL.parse(goSourceUrl);
 
       // TODO: `parsedUrl.pathname` can be undefined
-      const packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
+      let packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
 
-      const registryUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
+      const endpoint = GlobalConfig.get('endpoint');
+
+      const endpointPrefix = endpoint?.match('https://[^/]*/(.*)api/v4/?');
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:162]: `api/v4` should be optional, we will soon remove it from endpoint on platform init.

Comment (ManuelB) [reply lib/modules/datasource/go/base.ts]: @viceice I changed the regular expression and extended the test case. It is now optional.

**Comment by ManuelB** [comment]
@rarkins I now implemented this in the way that you requested in: https://github.com/renovatebot/renovate/discussions/26408

File: lib/modules/datasource/go/base.ts [hunk 2]
```diff
@@ -154,17 +155,35 @@ export class BaseGoDatasource {
...
+      const endpoint = GlobalConfig.get('endpoint');
+
+      const endpointPrefix = endpoint?.match('https://[^/]*/(.*?/)(api/v4/?)?');
+
+      if (endpointPrefix) {
+        packageName = packageName.replace(endpointPrefix[1], '');
+      }
+
+      const registryUrl =
+        endpoint?.replaceAll('api/v4/', '') ??
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:169]:
```suggestion
        endpoint?.replace(regEx('api/v4/?$'), '') ??
```

Should only occure once on the end.
I think the `registryUrl` should only be set to endpoint if `endpointPrefix` was matched.

File: lib/modules/datasource/go/base.ts [hunk 3]
```diff
@@ -154,17 +155,35 @@ export class BaseGoDatasource {
       const parsedUrl = URL.parse(goSourceUrl);
 
       // TODO: `parsedUrl.pathname` can be undefined
-      const packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
+      let packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
 
-      const registryUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
+      const endpoint = GlobalConfig.get('endpoint');
+
+      const endpointPrefix = endpoint?.match('https://[^/]*/(.*?/)(api/v4/?)?');
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:162]:
```suggestion
      const endpointPrefix = regEx('https://[^/]*/(.*?/)(api/v4/?)?'),exec(endpoint);
```

File: lib/modules/datasource/go/base.ts [hunk 4]
```diff
@@ -154,17 +155,35 @@ export class BaseGoDatasource {
       const parsedUrl = URL.parse(goSourceUrl);
 
       // TODO: `parsedUrl.pathname` can be undefined
-      const packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
+      let packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
 
-      const registryUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
+      const endpoint = GlobalConfig.get('endpoint');
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:160]:
```suggestion
      const endpoint = GlobalConfig.get('endpoint')!;
```
you can assume that `endpoint` isn't undefined here.

File: lib/modules/datasource/go/base.ts [hunk 5]
```diff
@@ -154,17 +155,35 @@ export class BaseGoDatasource {
       const parsedUrl = URL.parse(goSourceUrl);
 
       // TODO: `parsedUrl.pathname` can be undefined
-      const packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
+      let packageName = trimLeadingSlash(`${parsedUrl.pathname}`);
 
-      const registryUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
+      const endpoint = GlobalConfig.get('endpoint')!;
+
+      const endpointPrefix = regEx('https://[^/]*/(.*?/)(api/v4/?)?'),exec(endpoint);
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:162]:
```suggestion
      const endpointPrefix = regEx('https://[^/]*/(.*?/)(api/v4/?)?').exec(endpoint);
```

ups

Comment (ManuelB) [reply lib/modules/datasource/go/base.ts]: already recognized and fixed it :-)

**Comment by ManuelB** [comment]
@rarkins I fixed the prettier error:

```
vscode@6d06bfe1cec1:/workspaces/renovate$ git diff
diff --git a/lib/modules/datasource/go/base.spec.ts b/lib/modules/datasource/go/base.spec.ts
index b191d9d4c..e306ad6d1 100644
--- a/lib/modules/datasource/go/base.spec.ts
+++ b/lib/modules/datasource/go/base.spec.ts
@@ -412,7 +412,6 @@ describe('modules/datasource/go/base', () => {
           'example.com/gitlab/my-project/my-repo.git',
         );
 
-
         expect(res).toEqual({
           datasource: GitlabTagsDatasource.id,
           packageName: 'my-project/my-repo',
vscode@6d06bfe1cec1:/workspaces/renovate$ npx prettier --cache --check '**/*.{ts,js,mjs,json,md,yml}' "--cache-location" ".cache/prettier"
Checking formatting...
All matched files use Prettier code style!
```

File: lib/modules/datasource/go/base.spec.ts [hunk 6]
```diff
@@ -394,6 +395,49 @@ describe('modules/datasource/go/base', () => {
 
         expect(res).toBeNull();
       });
+
+      it('it correctly splits a URL where the endpoint is contained', async () => {
+        hostRules.hostType.mockReturnValue('gitlab');
+
+        GlobalConfig.set({ endpoint: 'https://example.com/gitlab/api/v4/' });
```
Comment (viceice) [inline lib/modules/datasource/go/base.spec.ts:402]: please call global config reset in before each callback

File: lib/modules/datasource/go/base.spec.ts [hunk 7]
```diff
@@ -394,6 +395,49 @@ describe('modules/datasource/go/base', () => {
...
+        );
+
+        expect(res2).toEqual({
+          datasource: GitlabTagsDatasource.id,
+          packageName: 'my-project/my-repo',
+          registryUrl: 'https://example.com/gitlab/',
+        });
+
+        GlobalConfig.reset();
+        hostRules.hostType.mockReturnValue('');
```
Comment (viceice) [inline lib/modules/datasource/go/base.spec.ts:439]: remove

File: lib/modules/datasource/go/base.ts [hunk 8]
```diff
@@ -154,17 +155,37 @@ export class BaseGoDatasource {
...
+        endpoint,
+      );
+
+      if (endpointPrefix) {
+        packageName = packageName.replace(endpointPrefix[1], '');
+      }
+
+      const registryUrl =
+        endpoint?.replace(regEx('api/v4/?$'), '') ??
+        `${parsedUrl.protocol}//${parsedUrl.host}`;
```
Comment (viceice) [inline lib/modules/datasource/go/base.ts:172]:
```suggestion
        endpoint prefix ? endpoint.replace(regEx('api/v4/?$'), '') :
        `${parsedUrl.protocol}//${parsedUrl.host}`;
```
only do the endpoint stuff when endpoint is a prefix of the current dep. endpoint is always set here.

File: lib/modules/datasource/go/base.spec.ts (hunk 6 above)
Comment (ManuelB) [reply lib/modules/datasource/go/base.spec.ts:404]: Done

File: lib/modules/datasource/go/base.ts (hunk 8 above)
Comment (ManuelB) [reply lib/modules/datasource/go/base.ts]: Done

**Comment by rarkins** [comment]
Yes, you can run `pnpm run lint-fix` locally to run all linting

**Comment by ManuelB** [comment]
```
$ pnpm run lint-fix
...
Linting: 239 file(s)
Summary: 0 error(s)
```

**Review by viceice — APPROVED** [review]

**Comment by ManuelB** [comment]
@viceice & @rarkins thank you guys. You rock! 🎉

**Pull request merged**